### PR TITLE
fix: update AiiDA version, docker, and post-init.sh

### PR DESCRIPTION
**Pull request by GeigerJ2**
# Fix and update the RenkuLab AiiDA template

This PR updates the AiiDA template for RenkuLab.

Regarding the broken profile creation on start-up, see PR #78.

## Main changes

- The `Dockerfile` is updated, and a new image pinned in `renku.ini` (using the latest `renku` 0.22 Python 3.10 base
  image and an AiiDA v2.5.1 installation)
- `aiida-activate` was removed as the [repo](https://github.com/chrisjsewell/activate_aiida) is outdated
- Instead, the AiiDA profile is set up via `verdi profile setup` in the `post-init.sh` script, where depending on the presence (absence) of the `archive_url` the `sqlite_zip` (`sqlite_dos`) backends are used
- Currently, the AiiDA Postgres backend (`psql_dos`) cannot be used due to missing `sudo` access &rarr; For inspection `sqlite_zip` and `sqlite_dos` should be sufficient (frankly, I tried resolving that using the `aiida-core` Docker images/Dockerfiles for some time, but didn't manage to make it work, and the effort might not be worth it if we stick to the other backends)
- RabbitMQ is working as expected as it is installed via the `aiida-core.services` `conda` package, the `consumer_timeout` variable is set, and the version warning disabled
- Minor changes to the `explore` notebook and the README

## Some further notes

Currently, the directory where the archive is downloaded is "repo", but one could also use the existing "data" directory instead? I think this ties in with our discussion about `verdi init`, @sphuber and @giovannipizzi. If we manage to merge these changes soon, we could update the setup here - for now it works in the classical way. Pinging @ltalirz and @khsrali here for info as well.

If somebody wants to take it for a spin before the merge, you can provide my [repository URL](https://github.com/GeigerJ2/contributed-project-templates) and the [latest commit SHA of this PR](https://github.com/SwissDataScienceCenter/contributed-project-templates/commit/c6ae7e013d5360d51fd0c4450d3bb695a34c1ec7) when [creating a new RenkuLab project](https://renkulab.io/projects/new), and fetch the updated template (see screenshot below).

Would be happy about feedback or suggestions for improvement. We also had the idea of providing some additional custom code, which already gives some interesting information about the archive (e.g. as a `pandas` dataframe), though, I think this should be as agnostic of the structure of the data as possible and would warrant a separate PR either way.

Lastly, when trying to import [this archive](https://archive.materialscloud.org/record/file?record_id=2100&file_id=18486bd2-4eaa-49aa-bd8d-eb88e02f394a&filename=eclab.aiida) **after** the notebook is created (that is, not providing an `archive_url` initially), I'm getting a `NotImplementedError` (see screenshot below), while pure inspection via `sqlite_zip` without importing works. I haven't investigated it further so far, but could this be possibly a case in which an automatic migration is required, @superstar54?

## Creating a project from the template

image

## `NotImplementedError` on `archive import`

image

### CODE REVIEW

**Comment by sphuber** [comment]
> Lastly, when trying to import [this archive](https://archive.materialscloud.org/record/file?record_id=2100&file_id=18486bd2-4eaa-49aa-bd8d-eb88e02f394a&filename=eclab.aiida) **after** the notebook is created (that is, not providing an `archive_url` initially), I'm getting a `NotImplementedError` (see screenshot below), while pure inspection via `sqlite_zip` without importing works. I haven't investigated it further so far, but could this be possibly a case in which an automatic migration is required, @superstar54?

This is expected. The `sqlite_zip` backend is read only. When you are importing, you are trying to add new data to the storage, which is not supported. When you create a profile using `sqlite_zip` for an existing archive, you are essentially just "mounting" it as a read-only storage, you are not adding anything to it. So you either create an empty `sqlite_dos` profile and _then_ import the archive, or you use `sqlite_zip` that just mounts the archive.

**Comment by GeigerJ2** [comment]
> This is expected. The sqlite_zip backend is read only.

When no `archive_url` is provided, the profile is initialized with `sqlite_dos`, though, to enable archive import further down the line. `jinja` gets evaluated when the RenkuLab project is created from the template, so I'm checking if `archive_url` is set and depending on that, the `post-init.sh` script is set up accordingly:

https://github.com/GeigerJ2/contributed-project-templates/blob/8dfb01bf9543a2d0a3268b2b8357c6135fe9536b/aiida/post-init.sh#L48-L60:
```python
{% else %}

# Without archive_url, generate profile using `core.sqlite_dos` backend
verdi profile setup core.sqlite_dos \
    --profile $aiida_profile \
    --first-name "$first_name" \
    --last-name "$last_name" \
    --email "$email" \
    --institution RenkuLab \
    --set-as-default \
    --non-interactive

{% endif %}
```

So the setup should be fine:
image

**Comment by GeigerJ2** [comment]
Just a ping if one of the maintainers can have a quick look here? Thanks!

**Comment by ltalirz** [comment]
I was in contact with @rokroskar when this started

**Comment by rokroskar** [comment]
Ah thanks for pinging me @ltalirz I missed this PR! Thank you @GeigerJ2 I'll have a look asap!

**Comment by rokroskar** [comment]
btw, for quickly creating a project with this PR, you can use [this link](https://renkulab.io/projects/new?data=eyJ1cmwiOiJodHRwczovL2dpdGh1Yi5jb20vR2VpZ2VySjIvY29udHJpYnV0ZWQtcHJvamVjdC10ZW1wbGF0ZXMiLCJyZWYiOiJmaXgvdXBkYXRlX2FpaWRhX3RlbXBsYXRlIn0%3D).

File: aiida/environment.yml
Comment (rokroskar) [inline aiida/environment.yml:9]: not needed anymore?

File: aiida/post-init.sh
Comment (rokroskar) [inline aiida/post-init.sh:5]: I think you should just combine the changes from #78 in here - unless there is a good reason not to do it?

File: aiida/requirements.txt
Comment (rokroskar) [inline aiida/requirements.txt:1]: add a newline please - alternatively, you could also put this in the `environment.yml` to keep all the dependencies in the same place.

**Review by rokroskar — CHANGES_REQUESTED** [review]
Thanks @GeigerJ2! Some minor comments. 

As mentioned in one of my comments - why not include the changes from #78 here?

**Comment by rokroskar** [comment]
btw there is also https://github.com/SwissDataScienceCenter/contributed-project-templates/pull/8 which would be great to either wrap up or close :)

**Comment by GeigerJ2** [comment]
Thanks, @ltalirz, for the ping to @rokroskar, and thanks to you for getting back quickly afterward. I'll take care of your comments here, and on #78 in the following days. This is just to confirm that I'm aware of them :)

File: aiida/post-init.sh
Comment (GeigerJ2) [reply aiida/post-init.sh:5]: I separated them because, in principle, the changes in #78 fixed the broken setup, even without updating the whole AiiDA setup (which was what we originally thought was required). Nonetheless, I went ahead and updated everything to keep up with the active AiiDA development since then. For me, it's also fine to just merge both changes into one PR, though.

File: aiida/requirements.txt
Comment (GeigerJ2) [reply aiida/requirements.txt:1]: Good point! I'll move it there.

File: aiida/environment.yml
Comment (GeigerJ2) [reply aiida/environment.yml:9]: I think this stems from some peculiarities of past versions, which I am not familiar with. Now, just installing
`aiida-core.services` and `aiida-core` into the conda environment should be sufficient. I'll try it out with an updated
docker image, and with the `aiida-core` dependency moved here.

File: aiida/post-init.sh
Comment (GeigerJ2) [reply aiida/post-init.sh:5]: One reason to separate it and merge the other one first is that we provide a direct link on the Materials Cloud Archive website to a prepopulated RenkuLab project, like here:
image

As the link points to this repo, it could be already fixed by just the merge of #78, even without the upgrading of the AiiDA version. That being said, I'd like to finalize this effort here soon, so if merging takes only a few more hours/days, we can also hold off and just merge this PR here.

Comment (rokroskar) [reply aiida/post-init.sh:5]: Yeah feel free to do it incrementally by all means. btw, I can send you the renku frog for a nicer icon in MC ;)

File: aiida/requirements.txt
Comment (GeigerJ2) [reply aiida/requirements.txt:1]: Not sure why, but `conda` installs AiiDA v2.2, despite the current default version on `conda-forge` being 2.5.1... Anyway, I pinned the version and everything seems to be fine, at least with the Docker image I created locally and pushed to the registry :)

Comment (rokroskar) [reply aiida/requirements.txt]: oh I meant you could still use `pip` but inside the `environment.yml` - either way is fine

Comment (GeigerJ2) [reply aiida/requirements.txt]: Alright, I think I figured out why it does that. v2.5.1 is not compatible with the base image due to the `erlang` version. I therefore pinned `aiida-core.services` to v2.2.2 (which would also be the default selected by `conda`), and am installing `aiida-core` at v2.5.1 via `pip` instead right now. Would be nice not to have this mixing, but it seems to work, and `renku` is also updated via `pip` down the line, so I guess it's the best setup for now :)
Pinning both versions should make it stable, at least for the foreseeable future.

File: aiida/post-init.sh
Comment (rokroskar) [inline aiida/post-init.sh:25]: This will add `repo_dir` to the `.gitignore` every time the session starts; it also makes the repo dirty which isn't so nice for the user. I think you can simply add `aiida_data` to the `.gitignore` and get rid of this line.

**Comment by GeigerJ2** [comment]
I addressed your changes in my recent commit. Along the way, I tried to restrict the installation to only using `conda`, and pin both `aiida-core` and `aiida-core.services` to the latest v2.5.1, but as seen in [my comment above](https://github.com/SwissDataScienceCenter/contributed-project-templates/pull/79#discussion_r1566160785) this was not straightforward. Another possible improvement I tried out was to install everything into a dedicated `aiida_renku` `conda` environment, rather than `base`, but the notebook wouldn't pick that up, so I reverted it. Using `base` should be fine for now.

> btw there is also https://github.com/SwissDataScienceCenter/contributed-project-templates/pull/8 which would be great to either wrap up or close :)

I don't have any opinion on that change. If nothing speaks against it, then I could add it here. 0.25 cpu seems indeed a bit small :D so you let me know. Thanks again, @rokroskar!

File: aiida/post-init.sh
Comment (GeigerJ2) [reply aiida/post-init.sh:5]: Feel free to share the frog with me and I'll send it to the MC people :D

File: aiida/post-init.sh
Comment (rokroskar) [inline aiida/post-init.sh:64]: This will break the next time `aiida-core` upgrades `rabbitmq` and will cause the container to crash. I believe you can do the same by setting an environment variable? See [here](https://stackoverflow.com/a/73928874/1065823). I suppose you could set that in the `Dockerfile`.

**Review by rokroskar — CHANGES_REQUESTED** [review]
Thanks @GeigerJ2! Just two more small things I noticed...

File: aiida/post-init.sh
Comment (GeigerJ2) [reply aiida/post-init.sh:25]: So, `post-init.sh` is run every time the session starts, not just when the project is created? I just verified that, and that's an issue, as the AiiDA profile should also only be created once. I could add a check for that in the `post-init.sh` file, but that seems a bit hacky. Is there any other place to put such setup steps that come after the docker image build, but are only run once on project creation, rather than at every session startup?

Regarding adding `aiida_data` to the `.gitignore`, good point, I'll add it there directly via its relative path, as you suggested.

Comment (sphuber) [reply aiida/post-init.sh:25]: > I just verified that, and that's an issue, as the AiiDA profile should also only be created once. I could add a check for that in the post-init.sh file, but that seems a bit hacky. Is there any other place to put such setup steps that come after the docker image build, but are only run once on project creation, rather than at every session startup?

It might not be a bad idea to still run this at each session startup in case the first time it failed for some reason. You can easily make the action idempotent by just checking if the target profile exists. Something like:
```
profile_name=some-profile
verdi profile show $profile_name 2> /dev/null || verdi profile setup ....
```

File: aiida/post-init.sh
Comment (GeigerJ2) [reply aiida/post-init.sh:64]: As both, the versions of `aiida-core` and `aiida-core.services` are pinned in the `Dockerfile`, it should not break, though? Anyway, I agree that this is something that can already be included during the build. Thanks for sending the link, I wasn't aware of that.

Btw, just out of curiosity, is there a recommended way to build the image? My current approach is to modify the `Dockerfile`, build it locally, and then push it to the registry, i.e.:
```shell
docker buildx build -t registry.renkulab.io/julian.geiger/aiida-renku-dev:latest ./aiida
docker push registry.renkulab.io/julian.geiger/aiida-renku-dev:latest
```
But for that to work, I'm always modifying this part:

```dockerfile
# RENKU_VERSION determines the version of the renku CLI
# that will be used in this image. To find the latest version,
# visit https://pypi.org/project/renku/#history.
ARG RENKU_VERSION={{ __renku_version__ | default("2.9.1") }}

# For local build
# ARG RENKU_VERSION="2.9.1"
```
to disable the `jinja` variable[^*], build the image, and enable it again before I commit. Building via the GitLab repo CI also seems a bit suboptimal, as I want the change to be applied to this repo, not the one I mirror on GitLab for the container registry. Thanks!

[^*]: Which gets filled in when the project is created, if I understand correctly?

Comment (rokroskar) [reply aiida/post-init.sh:64]: right, `aiida-core` is pinned, but there is no knowledge in the files here about which version of rabbitmq it bundles. So if you forget that this needs to be updated, it will fail. 

Regarding the image build - it's true that it isn't easy to build from the templated Dockerfile... I'm wondering if we could build the image in an action and push it, like we do [here](https://github.com/SwissDataScienceCenter/renkulab-docker/blob/main/.github/workflows/build-and-push-to-docker.yml) for our base images?

File: aiida/post-init.sh
Comment (GeigerJ2) [reply aiida/post-init.sh:25]: I'll resolve it like this, thanks, @sphuber!

File: aiida/post-init.sh
Comment (GeigerJ2) [reply aiida/post-init.sh]: `aiida-core.services` is pinned via `conda` so shouldn't that specify also the RMQ version (the `aiida-core` services are basically RMQ and PostgreSQL)? I'm not familiar with `conda` packaging so I might be wrong, though. In any case, I agree that modifying the config file in the hardcoded path is far from a desirable solution. I have changed it to use the ENV variable in the `Dockerfile`. Will give it a spin to see if it works and update here. Then we should hopefully be ready to merge.

Alright, at least I know I haven't been doing it in a wrong way. Yeah, an action would be preferrable I guess. For now, I think it's fine to leave it, but one could extend the [documentation](https://renku.readthedocs.io/en/0.51.0/reference/templates.html#pin-a-docker-image) in addition to the [Discourse post](https://renku.discourse.group/t/how-to-pin-a-docker-image/240). For me it was a bit confusing that I'm editing a repo from GitHub, but then the Docker image is hosted on the container registry of another project on GitLab via its CI/CD pipeline. So maybe just adding this option of local bulid and push to the container registry (once the project actually exists on GitLab) via, e.g. the two commands I posted used could be helpful :)

File: aiida/Dockerfile
Comment (rokroskar) [inline aiida/Dockerfile:6]: One thing that just occurred to me - it could be that you need to add this to the `~/.bashrc` rather than the container environment since RMQ is actually started from a subshell in the entrypoint?

Comment (GeigerJ2) [reply aiida/Dockerfile:6]: I just tested it with `rabbitmqctl eval 'application:get_env(rabbit, consumer_timeout).'` as suggested in the [Stackoverflow post you linked](https://stackoverflow.com/questions/70957962/rabbitmq-consumer-timeout-behavior-not-working-as-expected/73928874#73928874).

The behavior seems to be OK:

image

Haven't tested it by actually running AiiDA yet, though.

**Comment by GeigerJ2** [comment]
Thanks for looking into it @rokroskar! Is this ready to be merged, or are other changes required?

**Comment by GeigerJ2** [comment]
Pinging @rokroskar, as I'd like to close this soon. Thank you! :)

**Comment by rokroskar** [comment]
Apologies @GeigerJ2 I didn't get to this until now - I just tested the project and it looks good. I still think you might want to consider pinning the image you specify in `renku.ini` (i.e. not use `latest`) but you can also do this later if you want. Please resolve the conflict in the `Dockerfile` and then I can approve.

**Comment by GeigerJ2** [comment]
Hi @rokroskar, thanks for getting back here. I updated the branch via an interactive rebase, cleaning up the commit history a bit. I found the `latest` tag to be quite convenient, but I'm also happy with pinning it via the hash. Could you please remind me again which part of which hash I have to use here (checked [here](https://renku.readthedocs.io/en/stable/reference/templates.html#pin-a-docker-image) and [here](https://renku.discourse.group/t/how-to-pin-a-docker-image/240), but find the relevant information). I tried with the first and last seven characters of the commit hash on the [GitLab repo](https://gitlab.renkulab.io/julian.geiger/aiida-renku-dev/container_registry/25953) that hosts the image, as well as the image digest, but seems like these are not the correct specifications? Thanks!

**Comment by rokroskar** [comment]
hi @GeigerJ2 - you'll need to uncomment [these lines](https://gitlab.renkulab.io/julian.geiger/aiida-renku-dev/-/blob/master/.gitlab-ci.yml#L16-17) for the CI job to build the image.

**Comment by GeigerJ2** [comment]
> hi @GeigerJ2 - you'll need to uncomment [these lines](https://gitlab.renkulab.io/julian.geiger/aiida-renku-dev/-/blob/master/.gitlab-ci.yml#L16-17) for the CI job to build the image.

Ah, alright, I see. Though, the project on GitLab is now outdated compared to the state of the code here, as I just regularly built the image locally and pushed it to the registry. Is it fine for now to keep the `latest` tag and get this merged so that the link to Renku on Materials Cloud Archive works again? I'll update it properly once we release a new AiiDA version, as that will also contain useful new features for the Renku integration.

**Comment by rokroskar** [comment]
Yep, that's fine!

**Pull request merged**
